### PR TITLE
Add API to retrieve directories and file lists, in a cross-platform way.

### PR DESCRIPTION
**Pull request by CartBlanche**
### Description of Change ###

Given a Directory name,  respectively return either the sub-directories or files within that directory

### Bugs Fixed ###

- Related to issue https://github.com/xamarin/Essentials/issues/1551

### API Changes ###

Added: 

```csharp
public static partial class FileSystem {
    .
    .
    public string[] FileSystem.GetAppResourceDirectories(string path);
    public string[] FileSystem.GetAppResourceFiles(string path);
}
```

And associated cross-platform code.

### Behavioural Changes ###

None. New functionality.

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of `main` at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation ([see walkthrough](https://github.com/xamarin/Essentials/wiki/Documenting-your-code-with-mdoc))

### Sample UI Changes ###
**Android:**
Screenshot_1606994851

**iOS:**
Screenshot 2020-12-03 at 11 22 29

**Mac:**
Screenshot 2020-12-03 at 11 20 23

**Tizen:**
Tizen

**Windows.UWP:**
Windows

### CODE REVIEW

**Comment by CartBlanche** [comment]
I was unable to test the iOS Tester due to this provisioning blocker
Screenshot 2020-12-03 at 11 38 41

I believe My provisioning is correct as I've recently built and Code signed MacOS apps.

**Comment by CartBlanche** [comment]
Should I raise an issue about the theming in iOS Sample? This is in the screenshot above but I'm adding it here to highlight it here. Basically we are getting white text on a white background. Interestingly the 1st screen that lists all the Essentials seems to be using dark mode, but the next screens like this one, on iOS don't...
Screenshot 2020-12-03 at 11 22 29

**Comment by CartBlanche** [comment]
@mattleibow, @jamesmontemagno Finally borrowed my son's Windows machine to test this out and the sample, builds and runs on there too now. I think this is ready for you gents to take a look.

**Comment by CartBlanche** [comment]
Looks like a timeout failure on Android 29. While on MacOS Cake.Core is blowing up. 
If there is anything I can do to fix either of these error, let me know.

**Comment by CartBlanche** [comment]
@mattleibow Happy New Year! I was going to rebase this branch/PR, but seems I've had my permissions revoked. Does that mean I need to fork and submit PRs that way in future?

**Comment by mattleibow** [comment]
Ah, you might need to use a fork then. I don't think there is anything specific blocking, just haven't had a chance to test the idea in some corner cases that I don't know about yet. I'm trying to merge PRs now, so I'll be looking at this.

File: Xamarin.Essentials/FileSystem/FileSystem.android.cs [hunk 1]
```diff
@@ -53,6 +54,17 @@ static Task<Stream> PlatformOpenAppPackageFileAsync(string filename)
             }
         }
 
+        // Android does't have the concept for returning Directories (AFICT).
+        // As such, we only return string with no . in them. Not ideal but
+        // should work in most cases as Folders rarely contain dots
+        static string[] PlatformGetAppResourceDirectories(string path)
+            => Platform.AppContext.Assets.List(path).Where(file => !file.Contains(".")).Select(x => Path.Combine(path, x)).ToArray();
```
Comment (mattleibow) [inline Xamarin.Essentials/FileSystem/FileSystem.android.cs:61]: Can we wrap this for easier reading?

File: Xamarin.Essentials/FileSystem/FileSystem.android.cs [hunk 2]
```diff
@@ -53,6 +54,17 @@ static Task<Stream> PlatformOpenAppPackageFileAsync(string filename)
...
+
+        // This may also return Directories, but we can't restrict the list,
+        // as some files may not have an extention. So user is on their own from here.
+        static string[] PlatformGetAppResourceFiles(string path)
+            => Platform.AppContext.Assets.List(path).Select(x => Path.Combine(path, x)).ToArray();
```
Comment (mattleibow) [inline Xamarin.Essentials/FileSystem/FileSystem.android.cs:66]: Maybe wrap this too.

File: Xamarin.Essentials/FileSystem/FileSystem.android.cs [hunk 3]
```diff
@@ -53,6 +54,17 @@ static Task<Stream> PlatformOpenAppPackageFileAsync(string filename)
...
         }
 
+        // Android does't have the concept for returning Directories (AFICT).
+        // As such, we only return string with no . in them. Not ideal but
+        // should work in most cases as Folders rarely contain dots
+        static string[] PlatformGetAppResourceDirectories(string path)
+            => Platform.AppContext.Assets.List(path).Where(file => !file.Contains(".")).Select(x => Path.Combine(path, x)).ToArray();
+
+        // This may also return Directories, but we can't restrict the list,
+        // as some files may not have an extention. So user is on their own from here.
```
Comment (mattleibow) [inline Xamarin.Essentials/FileSystem/FileSystem.android.cs:64]: Is there no way to detect this? Not ending in a slash or something?

File: Xamarin.Essentials/FileSystem/FileSystem.ios.tvos.watchos.macos.cs [hunk 4]
```diff
@@ -38,6 +38,22 @@ static string GetDirectory(NSSearchPathDirectory directory)
             }
             return dirs[0];
         }
+
+        static string[] PlatformGetAppResourceDirectories(string path)
+        {
+            if (Directory.Exists(path))
```
Comment (mattleibow) [inline Xamarin.Essentials/FileSystem/FileSystem.ios.tvos.watchos.macos.cs:44]: What happens if I change my current directory? Does it always work, or maybe we need to first get the package directory and then append? What if I pass a rooted path, will that still be inside the app resources dir?

File: Xamarin.Essentials/FileSystem/FileSystem.uwp.cs [hunk 5]
```diff
@@ -24,6 +24,22 @@ static Task<Stream> PlatformOpenAppPackageFileAsync(string filename)
 
         internal static string NormalizePath(string path)
             => path.Replace('/', Path.DirectorySeparatorChar);
+
+        static string[] PlatformGetAppResourceDirectories(string path)
+        {
+            if (Directory.Exists(path))
```
Comment (mattleibow) [inline Xamarin.Essentials/FileSystem/FileSystem.uwp.cs:30]: Should this also be normalized?

Comment (mattleibow) [reply Xamarin.Essentials/FileSystem/FileSystem.uwp.cs:30]: Also, what happens if I pass a rooted path or change the cwd? Should we be checking for the location first?

File: Xamarin.Essentials/FileSystem/FileSystem.android.cs (hunk 1 above)
Comment (mattleibow) [reply Xamarin.Essentials/FileSystem/FileSystem.android.cs:61]: Android is dodgy and it seems if I pass in `/` then I might get the root of the APK or something based on SO. I tested with Android 30 and I get an empty result. Which is wrong. If you use `""`, then you get the root. Might be worth adding some checks to first make sure that the path coming in is translated to the Android ways.

However, Android is bananas (probably fitting since we support the mono monkey) because Android adds junk:

```
> Platform.AppContext.Assets.List("");
{string[4]}
	[0]: "FileSystemTemplate.txt"
	[1]: "Folder"
	[2]: "images"
	[3]: "webkit"
```

This guy person reports that is also has a `sounds` dir which I don't have: https://stackoverflow.com/a/8366081 However, we cannot filter any of them out since that really is a valid folder and I can add things to it... I can't see the folder in the APK, but I can add things to it and then it exists. Bonkers. It must be a pseudo file structure that is merged with the real APK.

Comment (mattleibow) [reply Xamarin.Essentials/FileSystem/FileSystem.android.cs:61]: I also don't think we can use the filter of "contains period" because I can create an extensionless file. Maybe if we switch to EnumerateFiles, we can also do a `Assets.List("<filename>")` and see if it returns anything:

```
var dirs = Assets.List(path);
foreach (var dir in dirs) 
    if (Assets.List(dir).Length > 0)
        yield return dir
    else
        // this is a file
```

And we can do the reverse for the EnumFiles

**Review by mattleibow — CHANGES_REQUESTED** [review]
This AIP is a bit ore complex based on things I mention below and in my comments. Let me know your thoughts and the things that didn't work for you.

1. The name of the method should maybe be `GetAppPackage[Files|Directories]` as the "package" idea is what the existing AIP uses.
2. We should probably work similarly to the `Directory` API does
   https://docs.microsoft.com/dotnet/api/system.io.directory.getfiles
   https://docs.microsoft.com/dotnet/api/system.io.directory.getdirectories
   1. Throw null exceptions for bad paths/incorrect paths/incorrect file/folder types.
   2. Throw IO exception the GetDirectories was a file
   3. Return an array of the full names (including paths) for the items in the specified directory, or an empty array if no items are found.
3. I notice that on some cases you assume the dev is a good guy and won't pass in rooted paths or change the working dir.
4. I would also suggest using the EnumerateFiles/Directories instead of Get and returning an `IEnumerable<string>`. If the result list is large, then there will be a significant delay. We can always have another method that does ToArray adjacent.

**Comment by mattleibow** [comment]
If you can't work on this branch, then just open a new PR and link from here.

**Comment by CartBlanche** [comment]
@mattleibow Thanks for the feedback. Looks like I may have to send a new PR from my fork. I'm currently in the middle of finishing and releasing my own app (hence the TTS PR), so won't get to resubmit a PR until next week. I definitely will revisit though, because as I mentioned this code is currently being used in a friend's language app on iOS, Android, Mac and they'll be releasing on  Windows by end of Feb.

**Comment by CartBlanche** [comment]
@mattleibow I'll close this PR as I've just created a draft PR from my fork. So future changes will be from there.
